include/chemfiles/shared_allocator.hpp

```diff
@@ -39,8 +42,9 @@ class shared_allocator {
     /// value of type T with the given arguments.
     template<class T, typename ... Args>
     static T* make_shared(Args&& ... args) {
+        auto instance = instance_.lock();
```

Comment:
Interesting, is this what fixed the bug?

Reply:
No, this is just to make the code more precise. The main fix here is moving `pointers_.erase(it);` out of the `if (count == 0)` block in free/release. A tangential fix is not using iterators after they have been deleted ...